### ISSUE TITLE
Chrome crashes in CI environments and is not re-launched

### ISSUE THREAD

**Issue by matanlurey** [dart-lang/test#724]
See https://github.com/dart-lang/angular/issues/705.

Note this only happens in more resource constrained environments (travis), so I can't reproduce it locally, but I can reproduce on travis almost every single time. I assume this has something to do with load.

**Comment by jakemac53** [comment]
Fwiw I don't think this has anything to do specifically with the --precompiled flag. Running large amounts of tests, specifically ddc tests, causes chrome to consume very large amounts of ram. On Travis it eventually dies as a result, locally it just gets really slow.

**Comment by matanlurey** [comment]
@jakemac53 noted this might be unrelated to `--precompiled`, just that with `j=1` (required for travis), and that Chrome gets hammered by hundreds of tests, and either hangs or crashes. I don't think we have any control here other than manual sharding :-/

**Comment by matanlurey** [comment]
This might be a memory leak in pkg/test? @jakemac53

**Comment by grouma** [comment]
Discussed offline. The theory is that chrome is crashing half way through the test suite. We should at least have better error handling here: https://cs.corp.google.com/piper///depot/google3/third_party/dart/test/lib/src/runner/browser/browser_manager.dart?rcl=174058812&l=108

**Comment by grouma** [comment]
I reran the tests with the new browser logging. No IO information was printed which leads me to believe the browser did not crash. It is likely just the tab crashed. See: https://travis-ci.org/dart-lang/angular/builds/310238302

I also ran the tests on Firefox. They all pass: https://travis-ci.org/dart-lang/angular/jobs/310243141

I will try running the tests again with piping the Chrome logs directly to the console regardless of crash.

**Comment by nex3** [comment]
Thanks for looking into this, Gary!

It might be worth trying to reproduce this locally by creating a directory full of test files that all do large allocations. If you can get a local repro, it'll probably be easier to iterate on than Travis.

**Comment by jakemac53** [comment]
Note that the firefox run was also _really_ slow (>25 mins), with the manual sharding on travis the chrome tests usually pass in more like 5 minutes. My guess is that it has the same memory leak issue but just does a better job at handling it than chrome (slows down but keeps going)?

It is also possible that ddc on firefox is simply slower but 5x seems a bit much (and if that is the case we should probably file some ddc bugs).

**Comment by matanlurey** [comment]
OK, I have a solid 100% reproduction, even locally:

Requires the latest `dev .20` SDK.

```bash
git clone https://github.com/dart-lang/angular
cd angular/_tests
pub run build_runner build -o build
```

... this will take a few minutes ...

```bash
pub run test --precompiled build -p chrome
```

Locally I get to ~700 or so, and then it will hang (747: `test/core/view/projection_integration_test.dart:` on my latest run). The timer continues to climb, but it never times out, succeeds, or fails. It's not the same test every-time, so I don't think it's related to a specific test.

**Comment by nex3** [comment]
Can you track the Chrome process's memory consumption and see if it's climbingprocess?

Does this continue to reproduce with `-j 1`? What about `-j 2`?

**Comment by matanlurey** [comment]
Trying `-j 1` now.

**Comment by matanlurey** [comment]
Just hangs at a different point for me:

`02:41 +753 ~2: compiling test/di/injector_test.dart`

**Comment by nex3** [comment]
When it hangs with `-j 1`, can you see anything unusual in Chrome's dev console?

**Comment by matanlurey** [comment]
How can I see the dev console, is there a way to run non-headless? @nex3

**Comment by nex3** [comment]
If you use `test` prior to version 0.12.30 it'll always run Chrome with a visible window.

**Comment by matanlurey** [comment]
I'd prefer to run in the latest version so I'm not debugging something that doesn't exist anymore or has changed/been refactored.

**Comment by nex3** [comment]
There haven't been any changes to browser infrastructure since 0.12.29+1, but if you'd prefer you can run it with `--pause-after-load` and manually tell it to continue before each suite

**Comment by nex3** [comment]
Looking into this locally, I notice two things:

1. Chrome eventually starts opening the debugger due to out-of-memory issues.
2. It looks like there are multiple iframes loaded at once when this happens, even with `-j 1`.

The latter is a bug regardless of anything else. I can't guarantee that fixing it will fix the crashes--this logic bug should presumably affect all browsers equally, so Chrome may still have a memory leak somewhere--but it's a first step.

**Comment by nex3** [comment]
Sorry, I spoke too soon--there aren't actually an unexpected number of iframes loaded (we just load 2x the amount of concurrency). It does look like this is a plain old memory leak.

**Comment by matanlurey** [comment]
Memory leak in package:test?

There very well might be a leak in our tests, but I'd imagine the tear down of every iframe should clean everything up?

**Comment by matanlurey** [comment]
Thanks for the update!

@jmesserly or @alan-knight or @vsmenon - could someone from the DDC encampment try to pair with @nex3 on this? Currently `pub run test --precompiled` crashes on medium-sized projects using DDC, potentially due to how DDC is bound in the browser, and not expecting many DDC apps to be loaded and unloaded?

**Comment by nex3** [comment]
It's possible that this may be related to https://bugs.chromium.org/p/chromium/issues/detail?id=800830 somehow.

**Comment by nex3** [comment]
I hacked `test` a bit to clean up everything and pause after running a bunch of suites, and it looks a lot like something browser-side is leaking memory. The heap is about 1300MB large, which is needless to say way too much for a page with essentially nothing going on. I see a bunch of closures like this:

![closures](https://user-images.githubusercontent.com/188/36009079-5cb09ddc-0cff-11e8-8a70-154a4ff50ae9.png)

with retaining paths like these:

![retaining-path-1](https://user-images.githubusercontent.com/188/36009109-8bc86a5a-0cff-11e8-9aa2-8f48672b9b9e.png)
![retaining-path-2](https://user-images.githubusercontent.com/188/36009108-8bb8d432-0cff-11e8-87a7-9740cc0e06cc.png)

I'm also seeing tons of `HTMLIFrameElement`s (despite having no `iframe`s on the page anymore) with retaining paths like this:

![retaining-path-3](https://user-images.githubusercontent.com/188/36009187-f120b4ac-0cff-11e8-9a2c-92bff0444681.png)

Note the `slow_template_instantiations_cache` in this last retaining path, which was also noted in https://github.com/emberjs/ember.js/issues/15872 which led to [the Chrome bug above](https://bugs.chromium.org/p/chromium/issues/detail?id=800830).

**Comment by nex3** [comment]
If we just didn't re-use instances at all, that would *drastically* effect overall performance. A better workaround might be to switch to a new tab after 30 test suites or so, but that would just be a band-aid on top of the underlying problem (and one that would take a non-trivial amount of engineering). If it's plausible there's something we can do in DDC to avoid this at a lower level, I think we should investigate that first.

**Comment by vsmenon** [comment]
It looks like the retained closures are the top-level module scripts.  It may be that they are being retained by requirejs and need to be unloaded between test runs: 

http://requirejs.org/docs/api.html#undef

**Comment by nex3** [comment]
It seems like requirejs *shouldn't* be able to do anything that can persist across iframes, but that may not mean it *doesn't*. I'll run an experiment on raw DDC without any angular code to see if this still reproduces.

**Comment by vsmenon** [comment]
In general, DDC should not persist across iframes either.

One potential caveat I can think of: if there is Dart code in the framework like: `window.parent.postMessage` that might be attaching something (inside `dart:html`) from the child frame's JS to the parent frame.

**Comment by nex3** [comment]
We do communicate with the child frame using `postMessage`, but I verified that we unregister the listeners from the parent frame and the child frame's listeners *shouldn't* persist after it's closed (although again that doesn't mean they *don't*).

**Comment by vsmenon** [comment]
I'm also worried about this line of code (despite the file name, this is used in DDC):

https://github.com/dart-lang/sdk/blob/master/tools/dom/src/dart2js_DOMImplementation.dart#L57

When a parent accesses the child window in Dart or vice versa, it's registering something (DDC relies on DOM monkey patching).

`postMessage` alone shouldn't need that.  If it's possible, can you try deleting the corresponding line in your `dart_sdk.js` and seeing if that helps?

**Comment by nex3** [comment]
I've created a bunch of empty test suites:

```dart
import 'package:test/test.dart';

void main() {
  test("success", () {});
}
```

and run a few experiments over them using my modified test runner that cleans up all the iframes after running 60 suites, but keeps the browser open. Here's what I've seen:

* Running these suites as-is on DDC leaks memory and ends up with ~1300MB allocated.
* Running them with `registerGlobalObject()` commented out behaves essentially the same as running them as-is.
* Running them compiled with dart2js instead does *not* leak memory and ends up with ~10MB allocated.
* Running them with DDC on Firefox ends up with ~142MB allocated. 
* Running them with dart2js on Firefox ends up with ~6MB allocated.

It looks like we're leaking memory on Firefox too, although at a low enough rate that it's way less of a problem than on Chrome. Here's the diff between the two Firefox snapshots:

![firefox-diff](https://user-images.githubusercontent.com/188/36051808-33429d5a-0da0-11e8-9a72-6e3fac485b02.png)

**Comment by nex3** [comment]
@vsmenon and I poked around, and it looks like what's going on here is that DDC is monkeypatching the parent window object in a way that persists after the iframe dies. Normally this wouldn't be allowed, but because the iframes are from the same origin as the host page, it is.

I've verified that if we force the iframes to come from a different origin, the page ends up with ~140MB allocated. It looks like there's still some sort of leak at work here, but it's vastly smaller than the existing one.

**Comment by nex3** [comment]
It looks like Firefox isn't cleaning up all the parent monkeypatches after the child iframes close, but it *is* nulling out references to the child frames' DOMs, which explains why it's leaking so much less memory than Chrome.

I'm wondering whether the leftover memory leak is related to require.js somehow. I'm going to see what I can see with different-origin frames.

**Comment by nex3** [comment]
I just re-ran the cross-origin experiment and now it looks like there is no extra memory leak. I'm not sure what I did differently (maybe I forgot to force a GC the first time?), but we may be in the clear if we can fix the prototype-munging.

I've filed https://github.com/dart-lang/sdk/issues/32113 to track the DDC leak. I think we can work around this really easily in `test` though by just calling `window.parent.postMessage()` via JS interop.
